### ISSUE TITLE
Nested objects from multiple tables left join

### ISSUE THREAD

**Issue by goncalo-oliveira** [vb-consulting/Norm.net#14]
The syntax for doing a join query with multiple tables is interesting, but it seems to only work with `INNER` joins. Let's consider we have an `Assets` and a `Vehicles` table, where there is an *optional* correspondence of 1-1. Here's what the query would look like (excluding grouping code for simplification).

```
var tuple = await connection.Cast()
    .ReadAsync<Asset, Vehicle?>(
        """
        select a.*, v.*
        from assets a
        left join vehicles v on a.id = v.id
        """
    )
    .ToListAsync();
```

If I change the query to an `INNER JOIN` it works and I kind of expected the `LEFT JOIN` to work also by setting a nullable in the tuple, but it doesn't.

If it's not a lot a work, I think this could be an interesting fix.

**Comment by vbilopav** [comment]
If I understand this correctly when you use `LEFT JOIN` in your query all values that are mapped to a `Vehicle` are null, and so:

- Current behavior - instances for Vehicle are created with all values set to `NULL`
- Expected behavior - instances for a `Vehicle` should be null, objects are not supposed to be created at all if all properties/values are null.

Is that correct?

Yes, that would be an interesting fix, but I'm not sure it would break something. Maybe I can add a new global configuration switch to support this behavior?

**Comment by vbilopav** [comment]
@goncalo-oliveira 

I just released a new version, package 5.3.2 is being indexed, here are the release notes:
https://github.com/vb-consulting/Norm.net/blob/master/changelog.md#532-2023-05-03

Closing this issue.

**Comment by goncalo-oliveira** [comment]
Hi @vbilopav. Sorry for my late reply, been quite busy.

> Expected behavior - instances for a Vehicle should be null, objects are not supposed to be created at all if all properties/values are null and if the type is nullable

Yes, this is true. Sometimes there isn't a vehicle for an asset and in these cases, the expectation is that item2 in the tuple is null.

I've downloaded the new release (5.3.2) and re-tested the scenario. It seems to be working fine at first glance, but this particular case I do have one vehicle per asset. I'll do some more tests and let you know if anything goes wrong.

Thank you for working on this!

**Comment by goncalo-oliveira** [comment]
Seems like it's still not working as expected. Unless there's a vehicle for every asset, it's not working for me.

```
Object reference not set to an instance of an object. System.NullReferenceException: Object reference not set to an instance of an object.    
at Norm.Mapper.NormExtensions.InvokeSetStruct[T,TProp](Delegate method, Boolean nullable, T instance, Object value)    
at Norm.Mapper.NormExtensions.InvokeSet[T](Delegate method, Boolean nullable, TypeCode code, T instance, Object value, Boolean isArray, StructType structType)    
at Norm.Mapper.NormExtensions.MapInstance[T](ReadOnlyMemory`1 tuple, T& instance, MapDescriptor& descriptor, ValueTuple`7[]& delegates)    
at Norm.Mapper.NormExtensions.Map[T1,T2](IAsyncEnumerable`1 tuples, Type type1, Type type2)+MoveNext()    
at Norm.Mapper.NormExtensions.Map[T1,T2](IAsyncEnumerable`1 tuples, Type type1, Type type2)+MoveNext()    
at Norm.Mapper.NormExtensions.Map[T1,T2](IAsyncEnumerable`1 tuples, Type type1, Type type2)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
```

**Comment by vbilopav** [comment]
I'll look into it ASAP (today).

Not sure exactly how to reproduce this. 

There are two unit tests that cover this:
https://github.com/vb-consulting/Norm.net/blob/master/Tests/PostgreSqlUnitTests/NullableInstancesUnitTests.cs
https://github.com/vb-consulting/Norm.net/blob/master/Tests/PostgreSqlSerialUnitTests/Configure_NullableInstances_Test.cs

Perhaps when values for that instance are sometimes null and sometimes not null.

**Comment by vbilopav** [comment]
@goncalo-oliveira I've added another test and it works fine: See the code below.

It basically tries to emulate a join operation where sometimes there are matches and sometimes not. Query:

```sql
select * from (
    values
        (1, 'a1', 11, 'b1'),
        (2, 'a2', null, null),
        (3, 'a3', 33, 'b3'),
        (4, 'a4', null, null)
) t(foo1, bar1, foo2, bar2)
```

That returns:
| foo1 | bar1 | foo2 | bar2 |
| ----- | ----- | ----- | ----- |
| 1 | a1 | 11 | b1 |
| 2 | a2 | `NULL` | `NULL` |
| 3 | a3 | 33 | b3 |
| 4 | a4 | `NULL` | `NULL` |


These tests work as expected. 

Can you please help me to provide a reproduction of your case so I can include that fix too? Thanks.


```csharp
class TestClass1
{
    public int? Foo1 { get; set; }
    public string? Bar1 { get; set; }
}

class TestClass2
{
    public int? Foo2 { get; set; }
    public string? Bar2 { get; set; }
}

NormOptions.Configure(options =>
{
    options.NullableInstances = true;
});

var result = connection
    .Read<TestClass1?, TestClass2?>(@"select * from (
        values
            (1, 'a1', 11, 'b1'),
            (2, 'a2', null, null),
            (3, 'a3', 33, 'b3'),
            (4, 'a4', null, null)
    ) t(foo1, bar1, foo2, bar2)")
    .ToList();

Assert.Equal(1, result[0].Item1?.Foo1);
Assert.Equal("a1", result[0].Item1?.Bar1);
Assert.Equal(11, result[0].Item2?.Foo2);
Assert.Equal("b1", result[0].Item2?.Bar2);

Assert.Equal(2, result[1].Item1?.Foo1);
Assert.Equal("a2", result[1].Item1?.Bar1);
Assert.Null(result[1].Item2?.Foo2);

Assert.Equal(3, result[2].Item1?.Foo1);
Assert.Equal("a3", result[2].Item1?.Bar1);
Assert.Equal(33, result[2].Item2?.Foo2);
Assert.Equal("b3", result[2].Item2?.Bar2);

Assert.Equal(4, result[3].Item1?.Foo1);
Assert.Equal("a4", result[3].Item1?.Bar1);
Assert.Null(result[3].Item2?.Foo2);
```

**Comment by goncalo-oliveira** [comment]
Hi @vbilopav. Thanks for working on this.

I downloaded your repository and did some tests with it directly, so that I could better understand what was going on. I did find out what was causing the issue and how to workaround it.

So although I am passing a *nullable* vehicle type

```csharp
var tuple = await connection.Cast()
            .ReadAsync<Asset, Vehicle?>( ... );
```

The underlying type is still *Vehicle* and an instance is still created even though all values are *NULL*. And it was by seeing exactly where it was crashing that I found out what was causing this; a *non-nullable* type in a property.

```csharp
public class Vehicle
{
    public int Id { get; set; }
   ...
}
```

This is what in reality causes the error. By setting that property as *nullable*, the error goes away when the value is *NULL*.

What I expected however, was that the entire *Vehicle* instance was *null* beforehand. But this might not be so easy to do, given your mapping strategy. I am happy with the workaround, now that I am aware of the reason, though it might be worth documenting this behaviour.
